File: frontend/src/api/endpoints/watches.ts
import { apiGet, apiPost, apiPatch, apiDelete } from '../client.ts';
import type { SocialDashboardWidget } from '../../features/studio/dashboard/types-social-dashboard.ts';

/** A Watch — agentic alerting (supersedes Alert). User-owned monitor over a
 *  Subject's scope_posts. See docs/alerts/watch-system-spec.md. */

export type SubjectMode = 'agents' | 'all_my_agents' | 'all_org_agents';
export type Grain = 'per_agent' | 'aggregate';

export interface Subject {
  mode: SubjectMode;
  agent_ids?: string[];
  grain?: Grain;
}

export type Reducer = 'count' | 'sum' | 'avg' | 'min' | 'max' | 'p50' | 'p90' | 'distinct';
export type Basis = 'absolute' | 'share' | 'change';
export type CompareOp = '>' | '>=' | '<' | '<=' | 'between';

export interface StructuredCondition {
  scope?: Record<string, unknown> | null;
  measure: { reducer: Reducer; field?: string | null };
  basis: Basis;
  share?: { denominator?: Record<string, unknown> | null } | null;
  change?: { vs: 'prior_window' } | null;
  group_by?: string | null;
  compare: { op: CompareOp; threshold: number; threshold2?: number | null };
}

export interface SemanticCondition {
  instruction: string;
  scope?: Record<string, unknown> | null;
}

export interface Trigger {
  kind: 'structured' | 'semantic';
  structured?: StructuredCondition | null;
  semantic?: SemanticCondition | null;
}

export interface Window {
  mode: 'cumulative' | 'rolling' | 'vs_prior';
  hours: number;
}

export type Channel = 'in_app' | 'email' | 'whatsapp';

export interface Action {
  tier: 'notify' | 'briefing';
  channels: Channel[];
  include_widgets?: boolean;
  recipients?: string[];
  /** Dashboard widgets rendered to PNGs in the email when include_widgets is on. */
  widgets?: SocialDashboardWidget[];
}

export interface WatchSource {
  kind: 'nl' | 'manual';
  nl_text?: string | null;
}

export interface Watch {
  watch_id: string;
  owner_uid: string;
  name: string;
  subject: Subject;
  trigger: Trigger;
  window: Window;
  eval_on: 'schedule' | 'run';
  eval_interval_sec?: number;
  action: Action;
  source: WatchSource;
  enabled: boolean;
  min_interval_sec?: number;
  created_at: string | null;
  updated_at: string | null;
  last_fired_at: string | null;
  trigger_count: number;
}

export interface WatchCreateBody {
  name: string;
  subject: Subject;
  trigger: Trigger;
  window?: Window;
  eval_on?: 'schedule' | 'run';
  eval_interval_sec?: number;
  action?: Action;
  source?: WatchSource;
  enabled?: boolean;
  min_interval_sec?: number;
}

export type WatchUpdateBody = Partial<{
  name: string;
  subject: Subject;
  trigger: Trigger;
  window: Window;
  eval_on: 'schedule' | 'run';
  eval_interval_sec: number;
  action: Action;
  enabled: boolean;
  min_interval_sec: number;
}>;

/** /watches/compile response: a reviewable draft or clarifying questions. */
export interface WatchCompileResult {
  status: 'watch' | 'clarification';
  draft?: WatchCreateBody;
  rationale?: string;
  clarifications?: string[];
}

export interface WatchPreviewResult {
  supported: boolean;
  reason?: string;
  would_fire?: boolean;
  value?: number | null;
  measure_label?: string;
  groups?: { key: string; value: number | null; fired: boolean }[];
  sample_post_ids?: string[];
  rows_scanned?: number;
}

export function listWatches(): Promise<{ watches: Watch[] }> {
  return apiGet('/watches');
}

export function createWatch(body: WatchCreateBody): Promise<Watch> {
  return apiPost('/watches', body);
}

export function compileWatch(nl_text: string, subject: Subject): Promise<WatchCompileResult> {
  return apiPost('/watches/compile', { nl_text, subject });
}

export function previewWatch(body: WatchCreateBody): Promise<WatchPreviewResult> {
  return apiPost('/watches/preview', body);
}

export function updateWatch(watchId: string, body: WatchUpdateBody): Promise<Watch> {
  return apiPatch(`/watches/${watchId}`, body);
}

export function deleteWatch(watchId: string): Promise<void> {
  return apiDelete(`/watches/${watchId}`);
}

/** A watch "covers" this agent if its subject names the agent or spans all agents. */
export function watchCoversAgent(w: Watch, agentId: string): boolean {
  if (w.subject.mode !== 'agents') return true;
  return (w.subject.agent_ids ?? []).includes(agentId);
}

export function summarizeTrigger(w: Watch): string {
  const t = w.trigger;
  if (t.kind === 'semantic') {
    return `AI watch: "${t.semantic?.instruction ?? ''}"`;
  }
  const s = t.structured;
  if (!s) return '—';
  const m = s.measure.reducer === 'count' ? 'count' : `${s.measure.reducer}(${s.measure.field})`;
  const core = s.basis === 'absolute' ? m : `${s.basis} of ${m}`;
  const gb = s.group_by ? ` per ${s.group_by}` : '';
  const thr = s.basis === 'share' ? `${Math.round(s.compare.threshold * 100)}%` : s.compare.threshold;
  return `${core}${gb} ${s.compare.op} ${thr}`;
}

File: api/schemas/watches.py
"""Watch — agentic alerting model (see docs/alerts/watch-system-spec.md, ADR-0005).

A Watch generalizes the legacy `Alert` (api/schemas/alerts.py): a user-owned
monitor that fires when a condition holds over a Subject's `scope_posts`. The
condition is one of two trigger kinds — `structured` (a deterministic query over
`scope_posts`) or `semantic` (a per-run LLM judge, phase 4). This module is the
persisted shape + request/response schemas; the detector that evaluates a
structured condition lives in `workers/watches/detector.py`.

Hard rules baked in here (from the design grill):
  * A Watch NEVER creates or mutates enrichment fields — `measure.field` may only
    reference fields that already exist on the agent (built-ins + custom).
  * `scope_posts` is the only metric source; daily_metrics/entity_metrics are not
    a dependency.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from api.routers.dashboard_schema import SocialDashboardWidget, SocialWidgetFilters

# Cap on dashboard widgets rendered into a watch's email (parity with the legacy
# Alert's MAX_WIDGETS_PER_ALERT). Defined here, not in the soon-deleted alerts.py.
MAX_WIDGETS_PER_WATCH = 4

# ── Subject: what a Watch monitors, resolved at eval time ───────────────────

SubjectMode = Literal["agents", "all_my_agents", "all_org_agents"]
Grain = Literal["per_agent", "aggregate"]


class Subject(BaseModel):
    model_config = ConfigDict(extra="ignore")

    mode: SubjectMode = "agents"
    # Only meaningful when mode == "agents". A single-agent watch is the
    # 1-element case; all_my_agents/all_org_agents resolve dynamically at eval.
    agent_ids: list[str] = Field(default_factory=list)
    grain: Grain = "per_agent"


# ── Structured trigger: a deterministic query over scope_posts ──────────────

Reducer = Literal["count", "sum", "avg", "min", "max", "p50", "p90", "distinct"]
# Built-in numeric fields plus `custom:<name>` and `custom:<name>.<element>` for
# list[object] element-grain. `distinct` pairs with a categorical field
# (channel_handle by default).
Basis = Literal["absolute", "share", "change"]
CompareOp = Literal[">", ">=", "<", "<=", "between"]


class Measure(BaseModel):
    model_config = ConfigDict(extra="ignore")

    reducer: Reducer = "count"
    # Ignored for reducer == "count". May be a built-in (views|likes|comments|
    # shares|saves|engagement_total), `custom:<name>`, or `custom:<name>.<elem>`.
    field: str | None = None


class ShareSpec(BaseModel):
    """basis == "share": numerator = the condition's `scope` rows; denominator =
    this filter applied to the same window (None → the whole agent scope)."""

    model_config = ConfigDict(extra="ignore")

    denominator: SocialWidgetFilters | None = None


class ChangeSpec(BaseModel):
    """basis == "change": measure(this window) / measure(prior window)."""

    model_config = ConfigDict(extra="ignore")

    vs: Literal["prior_window"] = "prior_window"


class Compare(BaseModel):
    model_config = ConfigDict(extra="ignore")

    op: CompareOp = ">"
    threshold: float
    threshold2: float | None = None  # only for op == "between"

    @field_validator("threshold2")
    @classmethod
    def _between_needs_two(cls, v, info):
        if info.data.get("op") == "between" and v is None:
            raise ValueError("op 'between' requires threshold2")
        return v


class StructuredCondition(BaseModel):
    model_config = ConfigDict(extra="ignore")

    # Sub-filter on the base scope_posts row-set (reuse dashboard vocab verbatim).
    scope: SocialWidgetFilters | None = None
    measure: Measure = Field(default_factory=Measure)
    basis: Basis = "absolute"
    share: ShareSpec | None = None
    change: ChangeSpec | None = None
    # Fires per-group, naming the culprit (e.g. per brand / per theme). One of the
    # dashboard dimension names (brands|themes|entities|platform|sentiment|...).
    group_by: str | None = None
    compare: Compare


class SemanticCondition(BaseModel):
    """Phase 4 — a per-run LLM judge over existing fields/content. Never compiles
    to a new enrichment field."""

    model_config = ConfigDict(extra="ignore")

    instruction: str = Field(min_length=1)
    scope: SocialWidgetFilters | None = None


# ── Window ──────────────────────────────────────────────────────────────────

WindowMode = Literal["cumulative", "rolling", "vs_prior"]


class Window(BaseModel):
    model_config = ConfigDict(extra="ignore")

    mode: WindowMode = "rolling"
    # Duration in hours for rolling / vs_prior; ignored for cumulative.
    hours: int = Field(default=168, ge=1)  # 7d default


# ── Action / delivery ───────────────────────────────────────────────────────

Channel = Literal["in_app", "email", "whatsapp"]
ActionTier = Literal["notify", "briefing"]


class Action(BaseModel):
    model_config = ConfigDict(extra="ignore")

    tier: ActionTier = "notify"
    channels: list[Channel] = Field(default_factory=lambda: ["in_app"])
    include_widgets: bool = False  # opt-in widget→PNG attachment, default off
    recipients: list[str] = Field(default_factory=list)  # default → owner email
    # Dashboard widgets rendered to PNGs and embedded in the email when
    # include_widgets is true (same schema the legacy Alert used → render path +
    # builder UI reused verbatim). Lives on Action so it serializes for free.
    widgets: list[SocialDashboardWidget] = Field(
        default_factory=list, max_length=MAX_WIDGETS_PER_WATCH
    )


# ── Source (NL provenance) ──────────────────────────────────────────────────


class WatchSource(BaseModel):
    model_config = ConfigDict(extra="ignore")

    kind: Literal["nl", "manual"] = "manual"
    nl_text: str | None = None  # source-of-truth for re-compile when kind == "nl"


# ── Trigger union + Watch create/update ─────────────────────────────────────

TriggerKind = Literal["structured", "semantic"]


class Trigger(BaseModel):
    model_config = ConfigDict(extra="ignore")

    kind: TriggerKind = "structured"
    structured: StructuredCondition | None = None
    semantic: SemanticCondition | None = None

    @field_validator("structured")
    @classmethod
    def _structured_present(cls, v, info):
        if info.data.get("kind") == "structured" and v is None:
            raise ValueError("structured trigger requires a structured condition")
        return v


class WatchCreate(BaseModel):
    model_config = ConfigDict(extra="ignore")

    name: str = Field(min_length=1, max_length=120)
    subject: Subject = Field(default_factory=Subject)
    trigger: Trigger = Field(default_factory=Trigger)
    window: Window = Field(default_factory=Window)
    eval_on: Literal["schedule", "run"] = "schedule"
    # How often the scheduler re-evaluates a schedule watch (advances next_eval_at).
    # Distinct from min_interval_sec, which throttles how often a STANDING-true
    # condition re-invokes the gate.
    eval_interval_sec: int = Field(default=3600, ge=300)
    action: Action = Field(default_factory=Action)
    source: WatchSource = Field(default_factory=WatchSource)
    enabled: bool = True
    # Anti-spam backstop only — not a suitability rule (see ADR-0005).
    min_interval_sec: int = Field(default=3600, ge=0)

    @field_validator("name")
    @classmethod
    def _strip_name(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("Watch name is required.")
        return v


class WatchUpdate(BaseModel):
    """Partial update — only provided keys are written."""

    model_config = ConfigDict(extra="ignore")

    name: str | None = Field(default=None, max_length=120)
    subject: Subject | None = None
    trigger: Trigger | None = None
    window: Window | None = None
    eval_on: Literal["schedule", "run"] | None = None
    eval_interval_sec: int | None = None
    action: Action | None = None
    enabled: bool | None = None
    min_interval_sec: int | None = None
